test/hasManyThroughUpsert.test.js

```diff
@@ -0,0 +1,179 @@
+var request = require("supertest");
+var loopback = require('loopback')
+var expect = require('chai').expect;
+var JSONAPIComponent = require('../')
+var RSVP = require('rsvp')
+
+var app
+var Movie, Category, MovieCategoryAssoc;
+
+describe('hasManyThrough upsert', function () {
+  beforeEach(function (done) {
+    app = loopback()
+    app.set('legacyExplorer', false)
+    var ds = loopback.createDataSource('memory')
+    // create models
+    Movie = ds.createModel('movie', {
+      id: {type: Number, id: true},
+      name: String
+    })
+
+    Category = ds.createModel('category', {
+      id: {type: Number, id: true},
+      name: String
+    })
+
+    MovieCategoryAssoc = ds.createModel('movieCategoryAssoc', {
+      id: {type: Number, id: true}
+    })
+
+    // add models
+    app.model(Movie)
+    app.model(Category)
+    app.model(MovieCategoryAssoc)
+
+    // set up relationships
+    Movie.hasMany(Category, { through: MovieCategoryAssoc })
+    Category.hasMany(Movie, { through: MovieCategoryAssoc })
+
+    MovieCategoryAssoc.belongsTo(Movie)
+    MovieCategoryAssoc.belongsTo(Category)
+    makeData().then(done())
```

Comment:
Yep, so two things

1. `makeData().then(done())` is basically equivalent to:

```
done()
makeData().then(undefined)
```

Which is not what you want.

2. If there is an error in makeData, it will not be picked up in the `.then` block. You need to add a catch block.

This line should be

```javascript
// makeData().then(done())
makeData().then(() => done()).catch(done)
```